crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -817,32 +823,54 @@ fn extract_end_of_epoch_tx(
         })
         .into_inner();
 
-    let digests = execution_digests
-        .last()
-        .expect("Final checkpoint must have at least one transaction");
+    // get a reverse iterator on execution digests
+    let mut rev_iter = execution_digests.iter().rev();
 
-    let change_epoch_tx = authority_store
-        .get_transaction_block(&digests.transaction)
-        .expect("read cannot fail");
+    let build_tx = |digests: &ExecutionDigests| {
+        let end_of_epoch_txn = authority_store
+            .get_transaction_block(&digests.transaction)
+            .expect("read cannot fail");
 
-    let change_epoch_tx = VerifiedExecutableTransaction::new_from_checkpoint(
-        change_epoch_tx.unwrap_or_else(||
-            panic!(
-                "state-sync should have ensured that transaction with digest {:?} exists for checkpoint: {checkpoint:?}",
-                digests.transaction,
-            )
-        ),
-        epoch_store.epoch(),
-        *checkpoint_sequence,
-    );
+        VerifiedExecutableTransaction::new_from_checkpoint(
+            end_of_epoch_txn.unwrap_or_else(||
+                panic!(
+                    "state-sync should have ensured that transaction with digest {:?} exists for checkpoint: {checkpoint:?}",
+                    digests.transaction,
+                )
+            ),
+            epoch_store.epoch(),
+            *checkpoint_sequence,
+        )
+    };
+
+    let last = rev_iter
+        .next()
+        .expect("Final checkpoint must have at least one transaction");
 
-    assert!(change_epoch_tx
+    let change_epoch_tx = build_tx(last);
+
+    assert!(&change_epoch_tx
         .data()
         .intent_message()
         .value
         .is_change_epoch_tx());
 
-    Some((*digests, change_epoch_tx))
+    end_of_epoch_txns.push((*last, change_epoch_tx));
+
+    // there may also be an AuthenticatorStateExpire transaction
+    if let Some(digests) = rev_iter.next() {
+        let tx = build_tx(digests);
+        if tx
+            .data()
+            .intent_message()
+            .value
+            .is_authenticator_state_expire_tx()
+        {
+            end_of_epoch_txns.push((*digests, tx));
+        }
+    }
```

Comment:
Does this need to be executed before the actual end of epoch txn? It looks to me like right now, given that this is pushed to the end for the vector, and then you call `execute_end_of_epoch_txn` twice serially, we will first enqueue  the actual end of epoch txn for execution, await its effects, then do the same for the authenticator state expiry txn.

If the two txns have a causal dependency, then it seems like you would end up deadlocking since they're in the same checkpoint and you enqueue the child and wait (and never get a chance to enqueue the parent).

If they're not, then don't we need to enforce that with the ordering here?

If order doesn't matter, then would it make sense to convert `execute_end_of_epoch_tx` to take a list, given that `handle_execution_effects` takes a list` so we could enqueue and await both in one shot